### DISCUSSION TITLE
Behavior of "not" keyword in the context of object pattern matching

### DISCUSSION THREAD

**Discussion by Crown0815** [dotnet/csharplang#4215] [Q&A] [answered]
I experimented a little with the `not` keyword in combination with the object pattern `{ }` which was introduced in C#8. I have come across some behavior that seems weird to me and I would like to make sure that this is intentional.

using something like
```csharp
if (obj is not { } nonNullObj) return)
nonNullObj.ToString(); // throws a null exception because nonNullObj is null. 
```

this is counter-intuitive to me since I would have expected 

```
obj is not { } nonNullObj
```

to be equivalent to

```
!(obj is { } nonNullObj)
```

which properly initialized `nonNullObj` to be not `null`. 

On top of that 

```csharp
if (new object is not { } nonNullObj) return)
nonNullObj.ToString();
```

actually works fine.

Maybe I am misunderstanding something here? I also opened a ticket on [StackOverflow](https://stackoverflow.com/questions/65252710/object-pattern-matching-results-in-uninitialized-variable) for this issue with some unit tests that reproduce the behavior.

**Comment by ufcpp** [comment, marked as answer]
Seems like already fixed bug https://github.com/dotnet/roslyn/issues/49708

**Comment by Crown0815** [comment]
@ufcpp Any idea when the fix will be released?

**Reply by 333fred** [reply]
With 16.9. I don't believe I have a public date that I can share for that.
